File: app/src/main/java/roboyard/eclabs/MapObjects.java
package roboyard.eclabs;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class MapObjects {

    // Compile patterns once as static fields for reuse
    private static final Pattern BOARD_SIZE_PATTERN = Pattern.compile("board:(\\d+),(\\d+);");
    private static final Pattern COORDINATE_PATTERN = Pattern.compile("(\\d+),(\\d+);");

    /*
     * Constructor of the class
     */
    public MapObjects(){

    }

    /*
     * Extract the data from the _data string
     * Returns a list of all extracted elements
     * Extracts the data from the _data string
     * Returns a list of all extracted elements
     * @return ArrayList<GridElement> List of grid elements
     */
    public static ArrayList<GridElement> extractDataFromString(String data)
    {
        int x = 0;
        int y = 0;

        // First check if the save contains board size information
        Matcher boardSizeMatcher = BOARD_SIZE_PATTERN.matcher(data);
        if (boardSizeMatcher.find()) {
            int boardX = Integer.parseInt(boardSizeMatcher.group(1));
            int boardY = Integer.parseInt(boardSizeMatcher.group(2));
            // Update and persist board size for this game
            MainActivity activity = GridGameScreen.gameManager.getActivity();
            activity.setBoardSize(activity, boardX, boardY);
        }

        ArrayList<GridElement> elements = new ArrayList<>();

        // r=robot (v=green, j=yellow, red, blue)
        // c=target
        // m=wall (horizontal, vertical)
        // v=vertical wall
        // h=horizontal wall
        List<String> objectTypes = Arrays.asList("mh", "mv", // wall (mur)
                "robot_green", "robot_yellow", "robot_red", "robot_blue", // robots
                "target_green", "target_yellow", "target_red", "target_blue", "target_multi"); // targets (cible)

        // Process each line of data once
        // Use negative limit to keep empty strings at end
        String[] lines = data.split(";", -1);
        for (String line : lines) {
            line = line.trim();
            if (line.isEmpty()) continue;
            
            // Find matching object type
            String matchedType = null;
            for (String type : objectTypes) {
                if (line.startsWith(type)) {
                    matchedType = type;
                    break;
                }
            }
            
            if (matchedType != null) {
                // Extract coordinates using indexOf instead of regex
                int coordStart = matchedType.length();
                String coords = line.substring(coordStart).trim();
                
                int commaIndex = coords.indexOf(',');
                if (commaIndex > 0) {
                    try {
                        String xStr = coords.substring(0, commaIndex).trim();
                        String yStr = coords.substring(commaIndex + 1).trim();
                        x = Integer.parseInt(xStr);
                        y = Integer.parseInt(yStr);
                        elements.add(new GridElement(x, y, matchedType));
                    } catch (NumberFormatException e) {
                    }
                } else {
                }
            } else {
            }
        }
        
        return elements;
    }

    /*
     * Generate a string containing all the information from the list
     * @param data List of GridElement containing all the content of the map
     * @param shortString Boolean if the string is squeezed into 5 letters
     * @return String containing all the map information like
     *   mv16,14;
     *   mv16,15;
     *   cj14,6;
     *   rr12,9;
     *   ...
     */
    public static String createStringFromList( ArrayList<GridElement> data, boolean shortString)
    {
        StringBuilder content = new StringBuilder();

        // Add board size information
        content.append("board:").append(MainActivity.boardSizeX).append(",").append(MainActivity.boardSizeY).append(";\n");

        // For each element, add a line containing the type as well as the x and y position
        for(GridElement currentElement : data)
        {
            content.append(currentElement.getType()).append(currentElement.getX()).append(",").append(currentElement.getY()).append(";\n");
        }

        String stringContent;
        if (shortString) {
            stringContent = generateUnique5LetterFromString(content.toString());
        }else{
            stringContent = content.toString();
        }

        return stringContent;
    }

    /*
     * set a color for the unique string depending on the value of the string,
     * the color is generated from the hex value of the string
     * - there are only generated light colors
     *
     * @param input The input string
     * @return A light color in hexadecimal format
     */
    public static String generateHexColorFromString(String input) {
        if (input == null || input.isEmpty()) {
            return "#000000";
        }
        
        // Simple and fast hash function
        int hash = 0;
        for (int i = 0; i < input.length(); i++) {
            hash = 31 * hash + input.charAt(i);
        }
        
        // Ensure colors are light by setting high base values
        int red = 128 + Math.abs(hash % 128);        // 128-255
        int green = 128 + Math.abs((hash >> 8) % 128);  // 128-255
        int blue = 128 + Math.abs((hash >> 16) % 128);  // 128-255
        
        return String.format("#%02X%02X%02X", red, green, blue);
    }

    /*
     * Generate a unique string from the input string
     * @param input The input string
     * @return A unique 5-letter string altering between vowels and consonants
     */
    public static String generateUnique5LetterFromString(String input) {
        try {
            // Create a SHA-256 message digest instance
            MessageDigest digest = MessageDigest.getInstance("SHA-256");

            // Get the hash bytes for the input string
            byte[] hashBytes = digest.digest(input.getBytes());

            // Define vowels and consonants
            char[] vowels = {'A', 'E', 'I', 'O', 'U'};
            char[] consonants = {'B', 'C', 'D', 'F', 'G', 'H', 'J', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'X', 'Y', 'Z'};

            // Convert the hash bytes to a 5-letter string, alternating between vowels and consonants
            StringBuilder uniqueString = new StringBuilder();
            for (int i = 0; i < 5; i++) {
                // Convert each byte to a positive integer and take modulo 26 to get a letter
                int index = Math.abs(hashBytes[i]) % (i % 2 == 0 ? consonants.length : vowels.length);
                // Map the index to an uppercase letter (ASCII code for 'A' is 65)
                char letter = (i % 2 == 0 ? consonants[index] : vowels[index]);
                // Append the letter to the unique string
                uniqueString.append(letter);
            }
            return uniqueString.toString();
        } catch (NoSuchAlgorithmException e) {
            // Handle NoSuchAlgorithmException if the specified algorithm is not available
            e.printStackTrace();
            return null; // or throw an exception
        }
    }
}

File: shared/src/commonMain/kotlin/roboyard/logic/core/MapObjects.kt
package roboyard.logic.core

import roboyard.logic.util.RLog


/**
 * Helper class for managing map objects and their properties.
 */
object MapObjects {
    private val log = RLog.tag("MapObjects")

    /**
     * Map object types for reference
     */
    const val TYPE_ROBOT: String = "robot"
    const val TYPE_TARGET: String = "target"
    const val TYPE_WALL: String = "wall"

    /**
     * Get the base type of a map object (robot, target, or wall)
     * @param objectType The full object type string (e.g., "robot_blue")
     * @return The base type string
     */
    fun getBaseType(objectType: String?): String {
        if (objectType == null) return "unknown"

        return if (objectType.startsWith("robot_")) {
            TYPE_ROBOT
        } else if (objectType.startsWith("target_")) {
            TYPE_TARGET
        } else if (objectType == "mh" || objectType == "mv") {
            TYPE_WALL
        } else {
            "unknown"
        }
    }

    /**
     * Get all available target types
     * @return List of target type strings
     */
    val allTargetTypes: MutableList<String>
        get() {
            val targets = mutableListOf<String>()
            targets.add("target_pink")
            targets.add("target_green")
            targets.add("target_blue")
            targets.add("target_yellow")
            targets.add("target_silver")
            targets.add("target_multi")
            return targets
        }

    /**
     * Get all available robot types
     * @return List of robot type strings
     */
    val allRobotTypes: MutableList<String>
        get() {
            val robots = mutableListOf<String>()
            robots.add("robot_pink")
            robots.add("robot_green")
            robots.add("robot_blue")
            robots.add("robot_yellow")
            robots.add("robot_silver")
            return robots
        }

    /**
     * Check if an object type is a robot
     * @param objectType The object type string
     * @return true if it's a robot
     */
    fun isRobot(objectType: String?): Boolean {
        return objectType != null && objectType.startsWith("robot_")
    }

    /**
     * Check if an object type is a target
     * @param objectType The object type string
     * @return true if it's a target
     */
    fun isTarget(objectType: String?): Boolean {
        return objectType != null && objectType.startsWith("target_")
    }

    /**
     * Check if an object type is a wall
     * @param objectType The object type string
     * @return true if it's a wall
     */
    fun isWall(objectType: String?): Boolean {
        return objectType == "mh" || objectType == "mv"
    }

    /**
     * Extract map data from a string
     */
    @JvmStatic
    fun extractDataFromString(data: String?): ArrayList<GridElement> = extractDataFromString(data, false)

    @JvmStatic
    fun extractDataFromString(data: String?, someFlag: Boolean): ArrayList<GridElement> {
        val result = ArrayList<GridElement>()
        if (data.isNullOrBlank()) return result

        // Compact code to full name mapping
        val compactCodeMap = mapOf(
            // Targets
            "tb" to "target_blue",
            "tr" to "target_red",
            "tg" to "target_green",
            "ty" to "target_yellow",
            "ts" to "target_silver",
            "tm" to "target_multi",
            "tp" to "target_pink",
            // Robots
            "rr" to "robot_red",
            "rb" to "robot_blue",
            "rg" to "robot_green",
            "ry" to "robot_yellow",
            "rs" to "robot_silver",
            "rp" to "robot_pink"
        )

        // Split data by semicolon
        val entries = data.split(";").filter { it.isNotBlank() }
        for (entry in entries) {
            // Very simple parsing for GridElement(x,y,type)
            // Expecting format like "mh1,2" or "robot_blue5,6" or "tb8,7"
            try {
                // Find where the numbers start
                val firstDigitIndex = entry.indexOfFirst { it.isDigit() }
                if (firstDigitIndex != -1) {
                    val type = entry.substring(0, firstDigitIndex)
                    // Convert compact code to full name if needed
                    val fullType = compactCodeMap[type] ?: type
                    val coords = entry.substring(firstDigitIndex).split(",")
                    if (coords.size == 2) {
                        result.add(GridElement(coords[0].toInt(), coords[1].toInt(), fullType))
                    }
                }
            } catch (e: Exception) {
                log.e("Error parsing map object data: $entry")
            }
        }

        return result
    }

    /**
     * Generate a unique 5-letter identifier from a string.
     * Deterministic - same input produces same output.
     * Alternates vowels and consonants.
     */
    @JvmStatic
    fun generateUnique5LetterFromString(input: String?): String {
        if (input.isNullOrEmpty()) return "aaaaa"
        
        val vowels = "aeiou"
        val consonants = "bcdfghjklmnpqrstvwxyz"
        
        // Simple hash of the input
        var hash = 0
        for (char in input) {
            hash = (hash * 31 + char.code) and 0x7FFFFFFF // Keep positive
        }
        
        // Generate 5 letters alternating consonant/vowel starting with consonant
        val result = StringBuilder()
        var useConsonant = true
        for (i in 0..4) {
            if (useConsonant) {
                result.append(consonants[hash % consonants.length])
                hash /= consonants.length
            } else {
                result.append(vowels[hash % vowels.length])
                hash /= vowels.length
            }
            useConsonant = !useConsonant
        }
        
        return result.toString()
    }
}
